Illya example sycl program using a cuda kernels static library

irudkin opened the pull request.

Commit:
1st version copy of the project in Codeplay GitLab 'HPC Illya Dev'. Not an exact copy due build include paths

irudkin committed:
Update README.md

Remove reference to NVIDIA Nsight.

irudkin committed:
Update README.md

Changed the readme's title and names of other projects this project mentions.

irudkin committed:
Update README.md

Changed the readme title.

rafbiels commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/.gitignore:
I would use wildcards here, the whole thing can be just:
```
*.o
*.a
StaticLibrarySYCLUsesCUDAKernels/bin/DXTC*
```
The first two could even go to a top-level `.gitignore` for the repository, but we can also do that later when we have more projects in the repo.

irudkin replied:
Done

rafbiels commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/.vscode/settings.json:
This file only contains your user preferences and is not relevant to the project, please remove it from the PR.

irudkin replied:
Done

rafbiels commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibraryBuildCUDAKernels/.vscode/c_cpp_properties.json:
This file contains user preferences specific to your machine/environment and is not relevant to the project, please remove it from the PR.

irudkin replied:
The VSCode setting is required for the VSCode project to operate. Without it a user would still need to set this up to work. If incorrect for their environment then they are free to edit it suit their needs by way of example.

rafbiels replied:
OK, let's keep it for now. Hopefully in future versions of the project we'll support all users independently of their choice of IDE thanks to standard CMake config, and we'll be able to remove all VSCode specific files.

rafbiels commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibraryBuildCUDAKernels/.vscode/launch.json:
There is no content in this file, can it be removed?

irudkin replied:
This is one of a standard set of vscode project files. Just happens there isn't a debugging going on here.

rafbiels replied:
OK, let's keep it for now. Hopefully in future versions of the project we'll support all users independently of their choice of IDE thanks to standard CMake config, and we'll be able to remove all VSCode specific files.

rafbiels commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibrarySYCLUsesCUDAKernels/.vscode/c_cpp_properties.json:
This file contains user preferences specific to your machine/environment and is not relevant to the project, please remove it from the PR.

irudkin replied:
The VSCode setting is required for the VSCode project to operate. Without it a user would still need to set this up to work. If incorrect for their environment then they are free to edit it suit their needs by way of example.

rafbiels replied:
OK, let's keep it for now. Hopefully in future versions of the project we'll support all users independently of their choice of IDE thanks to standard CMake config, and we'll be able to remove all VSCode specific files.

rafbiels commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibrarySYCLUsesCUDAKernels/.vscode/settings.json:
```diff
@@ -0,0 +1,84 @@
+{
+    "programName": "DXTC",
+    "files.associations": {
```
The `files.associations` setting is irrelevant to the project, please remove

irudkin replied:
These have been removed.

rafbiels replied:
The file seems to be still in the PR in the latest commit, it's just completely empty now. Can it be removed?

irudkin replied:
I cannot see a way to delete a file on the upstream repository. I need the file locally, so it cannot be deleted there. VSCode will 'edit' file (because it is its set of 4 required .cpp project files - they always get formed) when the user interacts with the project.      If I had ignored the file from the start this would not be a problem.  A way needs to be found to remove the file. Empty file is the next best thing as it will be filled out by VSCode.

rafbiels commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibraryBuildCUDAKernels/.vscode/settings.json:
```diff
@@ -0,0 +1,71 @@
+{
+    "programName": "StaticLibOfFnsCUDAKernelsDXTC",
+    "files.associations": {
```
The `files.associations` setting is irrelevant to the project, please remove

irudkin replied:
These have been removed.

rafbiels replied:
The file seems to be still in the PR in the latest commit, it's just completely empty now. Can it be removed?

irudkin replied:
I cannot see a way to delete a file on the upstream repository. I need the file locally, so it cannot be deleted there. VSCode will 'edit' file (because it is its set of 4 required .cpp project files - they always get formed) when the user interacts with the project.      If I had ignored the file from the start this would not be a problem.  A way needs to be found to remove the file. Empty file is the next best thing as it will be filled out by VSCode.

rafbiels reviewed (changes requested):
I left some comments about vscode settings that shouldn't be part of the project

Commit:
Simply the removal of temporary files

Commit:
Stop copy of VSCode preference setting

Commit:
remove stop copy of VSCode preference setting

irudkin committed:
Update settings.json

Remove VSCode personnel preference settings

irudkin committed:
Update settings.json

Remove VSCode personnel preferences

irudkin committed:
Update settings.json

Remove VSCode personnel preferences

Commit:
Stop the copy of VSCode personnel preference setting from all VSCode sub projects

Commit:
Fixing push
Merge remote-tracking branch 'refs/remotes/origin/Illya_SYCLuseCUDAKernelsStaticLibrary' into Illya_SYCLuseCUDAKernelsStaticLibrary

Commit:
Stop the copy of VSCode c_cpp_properties.json files

Commit:
Reinstate the inclusion of VSCode c_cpp_properties.json files being copied to upstream as there are setting required for the project to operate

Commit:
Removed settings.json from the project for other users (they will have their own)

Commit:
Removed settings.json from the project for other users (they will have their own)

Commit:
Added information about VSCode's configuration files for those not familiar with VSCode

Commit:
Bring back the VSCode settings.json file as it is required by the VScode build configuration

Commit:
Add to README.md guidance on how to keep VSCode settings from being overwritten locally

Commit:
Fix spelling mistake

DuncanMcBain commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/.gitignore:
```diff
@@ -0,0 +1,3 @@
+*.o
```
IMO it would make sense to put the object file ignore stuff at the top level. Github has a skeleton gitignore we could copy from for C++ projects

irudkin replied:
Is this a request for a change or just an opinion? We can certainly do this at a later date when the repository is more populated.

DuncanMcBain replied:
Personally I would add it here and now, but you don't have to.

irudkin replied:
A .gitignore file exists at one level up in codeexamples which has object file ignore rules included (along with other rules). Done.

DuncanMcBain commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/README.md:
```diff
@@ -0,0 +1,39 @@
+### VSCode and git
+As VSCode configuration files may have settings that are local to your development environment, you are likely to want to keep them, but not likely want them to be staged to be commited and so pushed up stream, nor have git overwrite your local settings.
+For the case of not wanting to push local changes, the git .gitignore file handles this. For case where local settings are to be kept you may want to do the following: 
+Perform a git stash, before getting the latest changes from the repository (by using git pull origin master or git rebase origin/master), and then merge your changes from the stash using git stash pop stash@{0}.
```
I'm not sure we need these instructions here (I don't think we're likely to take many pull requests from external collaborators).

irudkin replied:
Where would you like these instructions to be positioned in the repository if not here? ATM there will only a couple of examples. These instructions cannot be isolated from the context they are trying to inform the user.

irudkin replied:
I was not aware of the issue VSCode wanting local settings to a user but not wanting to put those setting in the main repository for everyone (would not suit everyone) and so this how to get around this problem. There is also this blog on the subject provided byRafel https://compiledsuccessfully.dev/git-skip-worktree/.

DuncanMcBain commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibraryBuildCUDAKernels/.vscode/c_cpp_properties.json:
```diff
@@ -0,0 +1,18 @@
+            "includePath": [
+                "${workspaceFolder}/**",
+                "${workspaceFolder}/../../cuda-samples-nvidia/Common",
+                "/usr/local/cuda-12.0/targets/x86_64-linux/include"
```
I don't think we should be hardcoding this path (for example, CUDA is definitely not installed there on my development machine).

irudkin replied:
As I said to Rafal, these configuration files are required for the project to work. Even if incorrect for a different user they can inform the user on how to edit the config to best for their environment.   Hard paths are never intentional, but there isn't an alternative I know to do this.

DuncanMcBain replied:
According to this documentation, https://code.visualstudio.com/docs/cpp/c-cpp-properties-schema-reference, this file is only used for Intellisense. It seems also that the plugin can fall back to default queries (to identify include paths and so on). But if it is only for intellisense, I'm not sure we even need to include it in the repository.

It can also work with `compile_commands.json` which is fully supported by CMake as well.

irudkin replied:
Yes required for Intellisense to operate correctly for the project. If Intellisense if not configured for a project, the VSCode IDE will display underscore squiggles through out the code to show it does not understand. This file is required.

DuncanMcBain commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibraryBuildCUDAKernels/.vscode/c_cpp_properties.json:
```diff
@@ -0,0 +1,18 @@
+                "/usr/local/cuda-12.0/targets/x86_64-linux/include"
+            ],
+            "defines": [],
+            "compilerPath": "/usr/local/cuda/bin/nvcc",
```
Similarly this isn't guaranteed to be the location of `nvcc`. Does `compilerPath` respect the `PATH` environment variable?

irudkin replied:
Same answer to the previous comment.

DuncanMcBain replied:
I've looked it up, compilerPath does not respect $PATH

irudkin replied:
So this file is the best it can be and being included does allow a user to edit the path to fit their needs by referencing an example. Information has been given in the readme for this project on the purpose of the (set of four) configuration files linking to a blog on how to set up VSCode for DPC++.

DuncanMcBain commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibraryBuildCUDAKernels/.vscode/launch.json:
```diff
@@ -0,0 +1,6 @@
+{
```
Do we need this empty file?

irudkin replied:
The file has been replaced with correct content. Done.

DuncanMcBain commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibraryBuildCUDAKernels/Makefile:
```diff
@@ -0,0 +1,15 @@
+build:
+	$(MAKE) -C bin all
+
+	
```
There's some whitespace here we can remove

irudkin replied:
White space removed from the file. Done.

DuncanMcBain commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibraryBuildCUDAKernels/bin/Makefile:
```diff
@@ -0,0 +1,28 @@
+	ranlib $(DST_FILE)_d.a
+
+build:
+	$(CUDA_PATH)/nvcc -rdc=true -c -ccbin g++ -I$(INCLUDE_PATH) -I/usr/local/cuda/bin/../targets/x86_64-linux/include -m64 --std=c++11 -gencode arch=compute_75,code=sm_75 ../src/$(SRC_FILE) -o $(DST_FILE)_tmp.o 
```
I don't think these hardcoded paths are always going to be correct, but even so we should at least normalise them

irudkin replied:
These paths are defined by the CUDA toolkit's installation, not by me, and so should be common to everyone relative a to root install directory. What do you mean by 'normalise them'?

DuncanMcBain replied:
In this context "normalise" means removing removing the relative components of the path, though I guess technically that might not actually be the same directory because of links. That being said this path is making a lot of assumptions anyway so it might be OK.

DuncanMcBain commented on codeexamples/SYCLUsingCUDAStaticLibraryOfKernels/StaticLibraryBuildCUDAKernels/src/library.cu:
```diff
@@ -0,0 +1,20 @@
+        return cudaGetLastError();
+    }
+
+} // namespace StaticLibFnsCUDAKernelsDXTC
```
This file is missing a newline at the end

irudkin replied:
I am not aware this is problem. What issue could this course?

DuncanMcBain replied:
It used to be disallowed for C++ source files. Now it is but it can be strange with tools that expect "normal" text files (which end with a newline followed by an EOF). Typically Microsoft tools haven't enforced it. If we had clang-format (which I guess we'll add later) then it would force the newlines at the end.

irudkin replied:
Change made. Done.

Commit:
Remove whitespace from Makefile

Commit:
Add to the readme about hardcoded paths and how the user should review and edit them to fit

Commit:
Add to the readme about Development environment

Commit:
Add end of file newline to ensure compatibility with some tools

Commit:
Replace the VSCode launch.json file with a non-empty one
